File: 27-XGBoost/xg_boost.R
#XGBoost


#import dataset
dataset = read.csv("Churn_Modelling.csv")
dataset = dataset[,4:14]

#transforming categorical data
dataset$Geography = as.numeric(factor(dataset$Geography,
                                      levels=c("France","Spain","Germany"),
                                      labels=c(1,2,3)))

dataset$Gender = as.numeric(factor(dataset$Gender,
                                   levels=c("Female","Male"),
                                   labels=c(1,2)))

#divide dataset into training and testing
library(caTools)
set.seed(123)
split = sample.split(dataset$Exited, SplitRatio = 0.80)
training_set = subset(dataset, split==TRUE)
testing_set = subset(dataset, split==FALSE)

#XGBOST DO NOT NEED SCALING VARIABLES


# ADJUST XGBOOST TO THE TRAINING
library(xgboost)
classifier = xgboost(data = as.matrix(training_set[,-11]),
                     label = training_set$Exited,
                     nrounds = 20)

#Kfold cross validations
library(caret)
folds = createFolds(training_set$Exited, k = 10)
cv = lapply(folds, function(x){
  training_fold = training_set[-x,]
  testing_fold = training_set[x,]
  classifier = xgboost(data = as.matrix(training_set[,-11]),
                       label = training_set$Exited,
                       nrounds = 20)
  y_pred = predict(classifier, newdata = as.matrix(testing_fold[,-11]))
  y_pred = (y_pred >= 0.5)#Esto se habla con la persona que desea el analisis para estimar un porcentaje
  #Crear la matriz de confusiones
  cm = table(testing_fold[,11], y_pred)
  accuracy = (cm[1,1]+cm[2,2])/(cm[1,1]+cm[1,2]+cm[2,1]+cm[2,2])
  return (accuracy)
})
mean(as.numeric(cv))
sd(as.numeric(cv))

File: 27-XGBoost/xg_boost.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr  9 22:09:50 2021

@author: Asus
"""
# =============================================================================
# XGBOOST
# =============================================================================

#libraries import
import numpy as np
import pandas as pd 

#data import
dataset = pd.read_csv('Churn_Modelling.csv')

#Matrix de caracteristicas
X = dataset.iloc[:,3:-1].values 
y = dataset.iloc[:,-1].values


#transforming categorical data
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

ctX = ColumnTransformer(
    [('one_hot_encoder', OneHotEncoder(categories='auto'), [1,2])],
    remainder='passthrough'                        
)
X = np.array(ctX.fit_transform(X), dtype=np.float)

# =============================================================================
# OJO Evita caer en la trampa de MULTICOLINEALIDAD para ambas columnas cambiadas
# =============================================================================
X = np.delete(X, (0,3), axis=1)


#divide dataset in training and testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# =============================================================================
# NO NEED TO scale variables in XGBOOST 
# =============================================================================
# =============================================================================
# #scaling variables
# from sklearn.preprocessing import StandardScaler
# sc_X = StandardScaler()
# X_train = sc_X.fit_transform(X_train)
# X_test = sc_X.transform(X_test)
# =============================================================================

# =============================================================================
# # BUILD XGBOOST TO TRAINING SET
# =============================================================================
import xgboost as xgb
classifier = xgb.XGBClassifier(learning_rate = 0.05, max_depth=3, n_estimators= 200, use_label_encoder=False)
classifier.fit(X_train, y_train)

# =============================================================================
# GRID SEARCH
# =============================================================================
from sklearn.model_selection import GridSearchCV
parameters = [{'max_depth':[3,6], 'n_estimators': [100,200],
               'learning_rate': [0.05, 0.1,0.2]}]

grid_search = GridSearchCV(estimator = classifier, 
                           param_grid= parameters,
                           scoring = 'accuracy',
                           cv = 20, # SAME AS KFOLD CORSS VAL
                           n_jobs = -1)

grid_search = grid_search.fit(X_train, y_train)

# =============================================================================
# GET THE BEST ACCURACY
# =============================================================================
best_accuracy = grid_search.best_score_

best_parameters = grid_search.best_params_


# =============================================================================
# #Predicciones
# =============================================================================
y_pred = classifier.predict(X_test)

#Elaborar matriz de confusion, para contrastar resultados de prediccion
from sklearn.metrics import confusion_matrix
c_matrix = confusion_matrix(y_true = y_test, y_pred = y_pred)

# K-fold cross validations
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10)
accuracies.mean()
accuracies.std()

